include/fx/gltf.h

```diff
@@ -1839,46 +1878,156 @@ namespace gltf
+        {
+            std::throw_with_nested(invalid_gltf_document("Invalid glTF document. See nested exception for details."));
+        }
+    }
+
+    inline Document LoadFromBinaryStream(std::istream & is, std::string const & bufferRootPath = "", ReadQuotas const & readQuotas = {})
+    {
+        try
+        {
+            detail::GLBHeader header;
+            std::size_t totalSize = 0;
+            std::vector<uint8_t> binary(detail::HeaderSize);
+            std::vector<uint8_t> extraChunks;
+            std::size_t extraChunksLength = 0;
+            {
+                is.read(reinterpret_cast<char *>(&binary[0]), detail::HeaderSize);
+                if (!is.good())
                 {
-                    throw std::system_error(std::make_error_code(std::errc::no_such_file_or_directory));
+                    throw std::system_error(std::make_error_code(std::errc::io_error));
                 }
-
-                const std::size_t fileSize = detail::GetFileSize(file);
-                if (fileSize < detail::HeaderSize)
+                std::memcpy(&header, &binary[0], detail::HeaderSize);
+                if (header.magic != detail::GLBHeaderMagic ||
+                    header.jsonHeader.chunkType != detail::GLBChunkJSON ||
+                    header.jsonHeader.chunkLength + detail::HeaderSize > header.length)
                 {
-                    throw invalid_gltf_document("Invalid GLB file");
+                    throw invalid_gltf_document("Invalid GLB header");
                 }
-
-                if (fileSize > readQuotas.MaxFileSize)
+                totalSize += detail::HeaderSize + header.jsonHeader.chunkLength;
+                if (totalSize > readQuotas.MaxFileSize)
                 {
                     throw invalid_gltf_document("Quota exceeded : file size > MaxFileSize");
                 }
-
-                binary.resize(fileSize);
-                file.read(reinterpret_cast<char *>(&binary[0]), fileSize);
+                binary.resize(totalSize);
+                is.read(reinterpret_cast<char *>(&binary[detail::HeaderSize]), header.jsonHeader.chunkLength);
+                if (!is.good())
+                {
+                    throw std::system_error(std::make_error_code(std::errc::io_error));
+                }
+                if (!is.eof())
+                {
+                    detail::ChunkHeader binHeader;
+                    is.read(reinterpret_cast<char *>(&binHeader), detail::ChunkHeaderSize);
+                    if (!is.good())
+                    {
+                        throw std::system_error(std::make_error_code(std::errc::io_error));
+                    }
+                    totalSize += detail::ChunkHeaderSize + binHeader.chunkLength;
+                    if (totalSize > readQuotas.MaxFileSize)
+                    {
+                        throw invalid_gltf_document("Quota exceeded : file size > MaxFileSize");
+                    }
+                    binary.resize(totalSize);

```

Comment:
This resize will also be pretty expensive...  Is it possible to read from the stream fewer times as well as make the sanity checks against the various sizes and limits just once?  This code path was already pretty tricky, fuzzing found some security issues, and now it's even more complex with all the offsets and additions.